### PR TITLE
[Serve] ensure replica reconfigure runs after allocation check

### PR DESCRIPTION
**Pull request by iasoon**
## Why are these changes needed?

Since remote calls provide no ordering guarantees, it could happen that `reconfigure` gets called before `is_allocated` Since `reconfigure` then runs the user initialization code, the replica actor could get blocked and never provide its allocation check.
This PR ensures that the allocation proof has been received before we run the replica initialization.

I believe this approach is better than handling the ordering in the calling code (in deployment_state), since that would require the controller to poll the replica in order to make progress on the initialization.

Would we want a test for this? What would that look like?
## Related issue number

https://github.com/ray-project/ray/pull/24044

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/serve/deployment_state.py [hunk 1]
```diff
@@ -336,7 +336,10 @@ def start(self, deployment_info: DeploymentInfo, version: DeploymentVersion):
 
         self._allocated_obj_ref = self._actor_handle.is_allocated.remote()
         self._ready_obj_ref = self._actor_handle.reconfigure.remote(
-            deployment_info.deployment_config.user_config
+            deployment_info.deployment_config.user_config,
+            # ensure that `reconfigure` will only be called after a response
+            # has been received from `is_allocated`.
```
Comment (edoakes) [inline python/ray/serve/deployment_state.py:341]:
```suggestion
            # Ensure that `reconfigure` will only be called after a response
            # has been received from `is_allocated`.
```
nit: capitalization

please also add *why* we care about the ordering here

File: python/ray/serve/replica.py [hunk 2]
```diff
@@ -201,8 +201,10 @@ async def is_allocated(self) -> str:
             return ray.get_runtime_context().node_id
 
         async def reconfigure(
-            self, user_config: Optional[Any] = None
+            self, user_config: Optional[Any] = None, _after: Optional[Any] = None
         ) -> Tuple[DeploymentConfig, DeploymentVersion]:
+            # unused `_after` argument is for scheduling: passing an ObjectRef
+            # allows delaying reconfiguration until after the object is ready.
```
Comment (edoakes) [inline python/ray/serve/replica.py:207]:
```suggestion
            # Unused `_after` argument is for scheduling: passing an ObjectRef
            # allows delaying reconfiguration until after this call has returned.
```

**Review by edoakes — APPROVED** [review]
Clever & simple solution! Disappointed that I didn't think of this...

Not sure how to write a test for this, but given the simplicity I don't think it's a hard requirement. Do you have any ideas for how it could be tested?

**Comment by edoakes** [comment]
@iasoon looks like there's a failure in the cross-language/java support:
https://buildkite.com/ray-project/ray-builders-pr/builds/29930#ce581916-8dda-4c06-a300-c52202e270a0/242-3751

@simon-mo could you provide some guidance on what to update to fix this?

**Comment by iasoon** [comment]
I think the failure is because I used a keyword argument for the reconfigure call, I think it's fixed now

File: python/ray/serve/deployment_state.py (hunk 1 above)
Comment (iasoon) [reply python/ray/serve/deployment_state.py]: Added a comment on that, but I find it a bit difficult to concisely explain. Feel free to nitpick it if you can think of something clearer!

Comment (edoakes) [reply python/ray/serve/deployment_state.py]: looks great 👍

**Review by simon-mo — APPROVED** [review]

**Pull request merged**